File: csharp/1320.minimum-distance-to-type-a-word-using-two-fingers.cs
/*
 * @lc app=leetcode id=1320 lang=csharp
 *
 * [1320] Minimum Distance to Type a Word Using Two Fingers
 */

using System;

// @lc code=start
public class Solution {

    private static int GetCost(int f, int t) {
        if (f == 26) {
            return 0;
        }
        return Math.Abs(f % 6 - t % 6) + Math.Abs(f / 6 - t / 6);
    }

    private static int MinimumDistanceRecurse(string word, int n, int i, int[,,] memo, int l, int r) {
        if (i >= n) {
            return 0;
        }
        if (memo[l, r, i] == 0) {
            int t = (int)word[i] - 'A';
            memo[l, r, i] = 1 + Math.Min(GetCost(l, t) + MinimumDistanceRecurse(word, n, i + 1, memo, t, r), GetCost(r, t) + MinimumDistanceRecurse(word, n, i + 1, memo, l, t));
        }
        return memo[l, r, i] - 1;
    }

    public int MinimumDistance(string word) {
        int n = word.Length;
        return MinimumDistanceRecurse(word, n, 0, new int[27, 27, n], 26, 26);
    }
}
// @lc code=end

File: java/1320.minimum-distance-to-type-a-word-using-two-fingers.java
/*
 * @lc app=leetcode id=1320 lang=java
 *
 * [1320] Minimum Distance to Type a Word Using Two Fingers
 */

// @lc code=start
class Solution {
    public int minimumDistance(String word) {
        return minimumDistanceRecurse(word, 0, 26, 26, new int[27][27][word.length()]);
    }

    private static int minimumDistanceRecurse(String word, int i, int l, int r, int[][][] memo) {
        if (i == word.length()) {
            return 0;
        }
        int v = memo[l][r][i];
        if (v == 0) {
            int c = word.charAt(i) - 'A';
            v = 1 + Math.min(getCost(l, c) + minimumDistanceRecurse(word, i + 1, c, r, memo), getCost(r, c) + minimumDistanceRecurse(word, i + 1, l, c, memo));
            memo[l][r][i] = v;
            memo[r][l][i] = v;
        }
        return v - 1;
    }

    private static int getCost(int a, int b) {
        if (a == 26) {
            return 0;
        }
        return Math.abs(a % 6 - b % 6) + Math.abs(a / 6 - b / 6);
    }
}
// @lc code=end

